pjsip/src/pjsip/sip_dialog.c

```diff
@@ -1727,7 +1728,18 @@ void pjsip_dlg_on_rx_request( pjsip_dialog *dlg, pjsip_rx_data *rdata )
     if (pjsip_rdata_get_tsx(rdata) == NULL &&
         rdata->msg_info.msg->line.req.method.id != PJSIP_ACK_METHOD)
     {
-        status = pjsip_tsx_create_uas(dlg->ua, rdata, &tsx);
+        /* Create tsx lock first so we can lock it before creating
+         * the transaction. This is to avoid deadlock by preventing
+         * the newly created transaction to process messages before
+         * this function returns.
+         */
+        status = pj_grp_lock_create(dlg->pool, NULL, &tsx_lock);
+        if (status == PJ_SUCCESS) {
+            pj_grp_lock_add_ref(tsx_lock);
+            pj_grp_lock_acquire(tsx_lock);
```

Comment:
minor: `pj_grp_lock_acquire()` will add ref too, so perhaps `pj_grp_lock_add_ref()` is not necessary?

Reply:
trying to acquire group lock with ref count 0 will trigger assertion:  `pj_assert(pj_atomic_get(glock->ref_cnt) > 0);`